### DISCUSSION TITLE
Configure to maintain book on Binance

### DISCUSSION THREAD

**Discussion by shaunc** [bmoscon/cryptofeed#809] [Q&A]
I am wondering how to configure the L2_BOOK feed so as to maintain a snapshot of the book with deltas and periodic snapshots when connecting to Binance. I see that `snapshot_interval` is listed in the 1.6.1 changelog, but it currently doesn't seem to be in the exchange codebase. I found this issue #604 -- but I am not familiar enough with the code to guess what the best practice is (until this is resolved?). Should I set `max_depth` to 5K .. 10K .. 20K? How do I control how often I receive a snapshot?

(NB this is for machine learning -- I do want the whole book or as much as I can get.)

Thank you!

**Comment by bmoscon** [comment]
the data object returned to the L2 callback has a delta field with the delta, and it has the book field with the complete orderbook. you can access either one at any time, so you can control yourself how often you take the snapshot vs the delta

**Reply by shaunc** [reply]
I'm not sure if I understand. I would assume that getting book snapshots at 100ms frequency is too resource intensive. (?) Are you saying that, internally, you poll for snapshots and maintain a "book" structure between snapshot calls using delta stream so that book snapshot reflects the current state? Are there any parameters to control this polling, or should we expect it to "just work"?

In particular the only parameter I see is "max depth" (and not polling interval). Are there circumstances in which one might want to increase beyond default 5K to get full book? ... or, after appropriate lead time to fill in sections of book not seen at first, is this default fine? Are there any other parameters?

**Reply by bmoscon** [reply]
the data comes from binance via a websocket, there is no polling. there is nothing you can do except to process these, there is no way to get them less frequently, unless you want to poll with the REST api. 

I assume someone might want to get a depth of 5k, so you can configure that if you want to.

**Reply by shaunc** [reply]
Hmm... deltas come from Binance via websocket. Complete order-book as well?? Is that a separate stream?

**Reply by shaunc** [reply]
Looking at the code, it seems to me that you do a [single retrieval of a snapshot when you receive first order book update](https://github.com/bmoscon/cryptofeed/blob/bc57694a31680681f64ba743c6451ebe096b7e4e/cryptofeed/exchanges/binance.py#L318) and that after that, you maintain the snapshot using the deltas. Do you ever refresh this snapshot?

Won't the problem be that in your initial poll, you'll miss some of the book, then when the price changes enough relevant deltas will stop making sense?

**Comment by bmoscon** [comment]
yep, the code follows the binance documentation. the documentation tells you to get a snapshot and the apply deltas to it. Is it possible that someone enters an order at the 5001th level and it eventually makes it into relevant levels above? Probably, but the binance documentation doesnt mention this, and it seems like an event of very low relevance to me. If you feel otherwise, please feel free to open a PR to address the issue.
